tests/unit_tests/dft/include/test_common.hpp

```diff
@@ -0,0 +1,228 @@
+/***************************************************************************
+*  Copyright (C) Codeplay Software Limited
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  For your convenience, a copy of the License has been included in this
+*  repository.
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+**************************************************************************/
+
+#ifndef ONEMKL_TEST_COMMON_HPP
+#define ONEMKL_TEST_COMMON_HPP
+
+#include <cstdint>
+#include <cstdlib>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+#if __has_include(<sycl/sycl.hpp>)
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
+
+template <typename T>
+struct complex_info {
+    using real_type = T;
+    static const bool is_complex = false;
+};
+
+template <typename T>
+struct complex_info<std::complex<T>> {
+    using real_type = T;
+    static const bool is_complex = true;
+};
+
+template <typename T>
+constexpr bool is_complex() {
+    return complex_info<T>::is_complex;
+}
+template <typename T>
+constexpr int num_components() {
+    return is_complex<T>() ? 2 : 1;
+}
+
+template <typename fp>
+typename std::enable_if<!std::is_integral<fp>::value, bool>::type check_equal(fp x, fp x_ref,
+                                                                              int error_mag) {
+    using fp_real = typename complex_info<fp>::real_type;
+    fp_real bound = (error_mag * num_components<fp>() * std::numeric_limits<fp_real>::epsilon());
+
+    bool ok;
+
+    fp_real aerr = std::abs(x - x_ref);
+    fp_real rerr = aerr / std::abs(x_ref);
+    ok = (rerr <= bound) || (aerr <= bound);
+    if (!ok) {
+        std::cout << "relative error = " << rerr << " absolute error = " << aerr
+                  << " limit = " << bound << std::endl;
+    }
+    return ok;
+}
+
+template <typename fp>
+bool check_equal(fp x, fp x_ref, int error_mag, std::ostream &out) {
+    bool good = check_equal(x, x_ref, error_mag);
+
+    if (!good) {
+        out << "Difference in result: DPC++ " << x << " vs. Reference " << x_ref << std::endl;
+    }
+    return good;
+}
```

Comment:
```suggestion
template <typename fp>
bool check_equal(fp x, fp x_ref, int error_mag) {
    return check_equal(x, x_ref, error_mag, std::cerr);
}
```
The use of an ostream is a little confusing here. The other overload should accept the ostream argument and this one should just call that with a default output stream.

Reply:
Can we also move the printing of the `x` vs `x_ref` to the other overload. (but don't call `x` "DPC++" since there will be multiple backends)

Reply:
Calling `x` *DPC++* is consistent with other pre-existing tests, so I think it should stay. Eg [in the BLAS tests](https://github.com/oneapi-src/oneMKL/blob/45c43edcfaea9773a01fcd32b1354984d6780b8e/tests/unit_tests/blas/include/test_common.hpp#L485).